// @aldridged/docusaurus-plugin-lunr/build/module/lib/env.js
import fs from 'fs-extra';
import path from 'path';
import { head, size } from 'lodash/fp';
import { VERSIONED_DOCS_DIR, VERSIONS_JSON_FILE } from './constants';
const VERSION_OPTIONS_DISABLED = {
    versioning: {
        docsDir: '',
        enabled: false,
        latestVersion: null,
        versions: [],
    }
};
function getVersionedDocsDir(siteDir) {
    return path.join(siteDir, VERSIONED_DOCS_DIR);
}
function getVersionsJSONFile(siteDir) {
    return path.join(siteDir, VERSIONS_JSON_FILE);
}
function getVersions(siteDir) {
    const versionsJSONFile = getVersionsJSONFile(siteDir);
    return fs.existsSync(versionsJSONFile) ? JSON.parse(fs.readFileSync(versionsJSONFile, 'utf8')) : [];
}
function getOptions(siteDir, versions) {
    return {
        versioning: {
            docsDir: getVersionedDocsDir(siteDir),
            enabled: true,
            latestVersion: head(versions),
            versions,
        }
    };
}
export default function (siteDir) {
    const versions = getVersions(siteDir);
    return size(versions) ? getOptions(siteDir, versions) : VERSION_OPTIONS_DISABLED;
}
