// @aldridged/docusaurus-plugin-lunr/build/module/lib/plugin.js
import fs from 'fs-extra';
import globby from 'globby';
import { flatten } from 'lodash/fp';
import lunr from 'lunr';
import path from 'path';
import loadEnv from './env';
import processMetadata from './metadata';
const DEFAULT_OPTIONS = {
    include: ['**/*.{md,mdx}'],
    path: 'docs',
    routeBasePath: 'docs',
};
export default function pluginContentLunr(context, opts) {
    const options = { ...DEFAULT_OPTIONS, ...opts };
    const { siteDir } = context;
    const docsDir = path.resolve(siteDir, options.path);
    // Versioning
    const env = loadEnv(siteDir);
    const { versioning } = env;
    const { versions, docsDir: versionedDir } = versioning;
    const versionsNames = versions.map(version => `version-${version}`);
    return {
        name: 'docusaurus-plugin-lunr',
        getThemePath() {
            return path.resolve(__dirname, '../theme');
        },
        // tslint:disable-next-line: readonly-array
        getPathsToWatch() {
            const { include } = options;
            const globPattern = include.map(pattern => `${docsDir}/${pattern}`);
            const versionGlobPattern = (!versioning.enabled) ? [] : flatten(include.map(p => versionsNames.map(v => `${versionedDir}/${v}/${p}`)));
            return [...globPattern, ...versionGlobPattern];
        },
        async loadContent() {
            const { include } = options;
            // tslint:disable-next-line: no-if-statement
            if (!fs.existsSync(docsDir)) {
                return null;
            }
            // Metadata for default/ master docs files.
            const docsFiles = await globby(include, { cwd: docsDir });
            const docsPromises = docsFiles.map(async (source) => processMetadata({
                context,
                env,
                options,
                refDir: docsDir,
                source,
            }));
            // Metadata for versioned docs
            const versionedGlob = flatten(include.map(p => versionsNames.map(v => `${v}/${p}`)));
            const versionedFiles = await globby(versionedGlob, { cwd: versionedDir });
            const versionPromises = (!versioning.enabled) ? [] : versionedFiles.map(async (source) => processMetadata({
                context,
                env,
                options,
                refDir: versionedDir,
                source,
            }));
            const metadata = await Promise.all([...docsPromises, ...versionPromises]);
            return ({ metadata });
        },
        contentLoaded({ content, actions }) {
            const { metadata = [] } = content;
            const { createData } = actions;
            // tslint:disable: no-expression-statement no-this typedef
            const index = lunr(function () {
                this.ref('route');
                this.field('title');
                this.field('content');
                this.field('version');
                metadata.forEach(function ({ permalink, title, version, plaintext }) {
                    this.add({
                        content: plaintext,
                        route: permalink,
                        title,
                        version
                    });
                }, this);
                // tslint:enable: no-expression-statement no-this typedef
            });
            const documents = metadata.map(({ permalink: route, title, version }) => ({ route, title, version }));
            // tslint:disable-next-line: no-expression-statement
            createData('search-index.json', JSON.stringify({ index, documents }, null, 2));
        }
    };
}
;
