// @aldridged/docusaurus-plugin-lunr/build/main/lib/env.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
const fp_1 = require("lodash/fp");
const constants_1 = require("./constants");
const VERSION_OPTIONS_DISABLED = {
    versioning: {
        docsDir: '',
        enabled: false,
        latestVersion: null,
        versions: [],
    }
};
function getVersionedDocsDir(siteDir) {
    return path_1.default.join(siteDir, constants_1.VERSIONED_DOCS_DIR);
}
function getVersionsJSONFile(siteDir) {
    return path_1.default.join(siteDir, constants_1.VERSIONS_JSON_FILE);
}
function getVersions(siteDir) {
    const versionsJSONFile = getVersionsJSONFile(siteDir);
    return fs_extra_1.default.existsSync(versionsJSONFile) ? JSON.parse(fs_extra_1.default.readFileSync(versionsJSONFile, 'utf8')) : [];
}
function getOptions(siteDir, versions) {
    return {
        versioning: {
            docsDir: getVersionedDocsDir(siteDir),
            enabled: true,
            latestVersion: fp_1.head(versions),
            versions,
        }
    };
}
function default_1(siteDir) {
    const versions = getVersions(siteDir);
    return fp_1.size(versions) ? getOptions(siteDir, versions) : VERSION_OPTIONS_DISABLED;
}
exports.default = default_1;
