// @aldridged/docusaurus-plugin-lunr/build/main/lib/plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = __importDefault(require("fs-extra"));
const globby_1 = __importDefault(require("globby"));
const fp_1 = require("lodash/fp");
const lunr_1 = __importDefault(require("lunr"));
const path_1 = __importDefault(require("path"));
const env_1 = __importDefault(require("./env"));
const metadata_1 = __importDefault(require("./metadata"));
const DEFAULT_OPTIONS = {
    include: ['**/*.{md,mdx}'],
    path: 'docs',
    routeBasePath: 'docs',
};
function pluginContentLunr(context, opts) {
    const options = Object.assign(Object.assign({}, DEFAULT_OPTIONS), opts);
    const { siteDir } = context;
    const docsDir = path_1.default.resolve(siteDir, options.path);
    // Versioning
    const env = env_1.default(siteDir);
    const { versioning } = env;
    const { versions, docsDir: versionedDir } = versioning;
    const versionsNames = versions.map(version => `version-${version}`);
    return {
        name: 'docusaurus-plugin-lunr',
        getThemePath() {
            return path_1.default.resolve(__dirname, '../theme');
        },
        // tslint:disable-next-line: readonly-array
        getPathsToWatch() {
            const { include } = options;
            const globPattern = include.map(pattern => `${docsDir}/${pattern}`);
            const versionGlobPattern = (!versioning.enabled) ? [] : fp_1.flatten(include.map(p => versionsNames.map(v => `${versionedDir}/${v}/${p}`)));
            return [...globPattern, ...versionGlobPattern];
        },
        async loadContent() {
            const { include } = options;
            // tslint:disable-next-line: no-if-statement
            if (!fs_extra_1.default.existsSync(docsDir)) {
                return null;
            }
            // Metadata for default/ master docs files.
            const docsFiles = await globby_1.default(include, { cwd: docsDir });
            const docsPromises = docsFiles.map(async (source) => metadata_1.default({
                context,
                env,
                options,
                refDir: docsDir,
                source,
            }));
            // Metadata for versioned docs
            const versionedGlob = fp_1.flatten(include.map(p => versionsNames.map(v => `${v}/${p}`)));
            const versionedFiles = await globby_1.default(versionedGlob, { cwd: versionedDir });
            const versionPromises = (!versioning.enabled) ? [] : versionedFiles.map(async (source) => metadata_1.default({
                context,
                env,
                options,
                refDir: versionedDir,
                source,
            }));
            const metadata = await Promise.all([...docsPromises, ...versionPromises]);
            return ({ metadata });
        },
        contentLoaded({ content, actions }) {
            const { metadata = [] } = content;
            const { createData } = actions;
            // tslint:disable: no-expression-statement no-this typedef
            const index = lunr_1.default(function () {
                this.ref('route');
                this.field('title');
                this.field('content');
                this.field('version');
                metadata.forEach(function ({ permalink, title, version, plaintext }) {
                    this.add({
                        content: plaintext,
                        route: permalink,
                        title,
                        version
                    });
                }, this);
                // tslint:enable: no-expression-statement no-this typedef
            });
            const documents = metadata.map(({ permalink: route, title, version }) => ({ route, title, version }));
            // tslint:disable-next-line: no-expression-statement
            createData('search-index.json', JSON.stringify({ index, documents }, null, 2));
        }
    };
}
exports.default = pluginContentLunr;
;
